apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableMutateRowsRequestManager.cs

```diff
@@ -0,0 +1,205 @@
+        {
+            /// <summary> All responses produced OK or a retryable code. </summary>
+            Retryable,
+
+            /// <summary> At least one response had a non-retryable code. </summary>
+            NotRetryable,
+
+            /// <summary> The response was invalid, missing index, etc. </summary>
+            Invalid
+        }
+
+        /// <summary>
+        /// Constructor for <see cref="BigtableMutateRowsRequestManager"/>.
+        /// </summary>
+        /// <param name="retryStatuses">
+        /// Collection of Grpc status codes to retry on.</param>
+        /// <param name="mutateRowsRequest">
+        /// <see cref="MutateRowsRequest"/> that was received from the user.</param>
+        internal BigtableMutateRowsRequestManager(IEnumerable<StatusCode> retryStatuses, MutateRowsRequest mutateRowsRequest)
+        {
+            _originalRequest = mutateRowsRequest;
+
+            _results = new Rpc.Status[_originalRequest.Entries.Count];
+
+            // This is a map between RetryRequest and _originalRequest
+            // as they could diverge if a retry occurs.
+            _mapToOriginalIndex = Enumerable.Range(0, _originalRequest.Entries.Count).ToArray();
+            
+            _rertryableCodes = new HashSet<int>(retryStatuses.Select(status => (int)status));
+        }
+
+        /// <summary>
+        /// Reads MutateRowsResponse message and saves the statuses of all request Entries.
+        /// </summary>
+        /// <param name="response">
+        /// A MutateRowsResponse message received from MutateRows call.</param>
+        internal void SetStatus(MutateRowsResponse response)
+        {
+            foreach (var entry in response.Entries)
+            {
+                int index = (int)entry.Index;
+
+                if (index >= _mapToOriginalIndex.Length || index < 0)
+                {
+                    _messageIsInvalid = true;
+                    break;
+                }
+                // Set the result
+                _results[_mapToOriginalIndex[index]] = entry.Status;
+            }
+        }
+
+        /// <summary>
+        /// This method is called when all calls to <see cref="SetStatus(MutateRowsResponse)"/> are complete.
+        /// </summary>
+        /// <returns>
+        /// ProcessingStatus of the accumulated responses - invalid, retryable, non-retryable.
+        /// </returns>
+        internal ProcessingStatus OnOk()
+        {
+            // Sanity check to make sure that every mutation received a response.
+            _messageIsInvalid = _messageIsInvalid || _results.Any(result => result == null);
+            
+            // There was a problem in the data found in SetStatus(), so fail the RPC.
+            if (_messageIsInvalid)
+            {
+                return ProcessingStatus.Invalid;
+            }
+
+            // This list hold indices of entries to be included in retry request
+            List<int> toRetry = new List<int>();
```

Comment:
Similarly to avoid heap allocations when not necessary, initialize this to null here and the create it if necessary before the first `Add`.